guacamole-ext/src/main/java/org/apache/guacamole/properties/CipherGuacamoleProperty.java

```diff
@@ -0,0 +1,92 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.guacamole.properties;
+
+import java.io.BufferedInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.InputStream;
+import java.io.IOException;
+import java.lang.IllegalArgumentException;
+import java.security.InvalidKeyException;
+import java.security.KeyFactory;
+import java.security.NoSuchAlgorithmException;
+import java.security.PrivateKey;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.KeySpec;
+import java.security.spec.PKCS8EncodedKeySpec;
+import javax.crypto.Cipher;
+import javax.crypto.NoSuchPaddingException;
+import org.apache.guacamole.GuacamoleException;
+import org.apache.guacamole.environment.Environment;
+import org.apache.guacamole.environment.LocalEnvironment;
+
+/**
+ * A GuacamoleProperty whose value is derived from a private key file.
+ */
+public abstract class CipherGuacamoleProperty implements GuacamoleProperty<Cipher>  {
+
+    @Override
+    public Cipher parseValue(String value) throws GuacamoleException {
+
+        try {
+
+            final Environment environment = new LocalEnvironment();
+
+            // Open and read the file specified in the configuration.
+            File keyFile = new File(environment.getGuacamoleHome(), value);
+            InputStream keyInput = new BufferedInputStream(new FileInputStream(keyFile));
+            final byte[] keyBytes = new byte[(int) keyFile.length()];
+            keyInput.read(keyBytes);
+            keyInput.close();
+
+            // Set up decryption infrastructure
+            KeyFactory keyFactory = KeyFactory.getInstance("RSA");
+            KeySpec keySpec = new PKCS8EncodedKeySpec(keyBytes);
+            final PrivateKey privateKey = keyFactory.generatePrivate(keySpec);
+            final Cipher cipher = Cipher.getInstance(privateKey.getAlgorithm());
+            cipher.init(Cipher.DECRYPT_MODE, privateKey);
+
+            return cipher;
+
+        }
+        catch (FileNotFoundException e) {
+            throw new GuacamoleException("Could not find the specified key file.", e);
+        }
+        catch (IOException e) {
+            throw new GuacamoleException("Could not read in the specified key file.", e);
+        }
+        catch (NoSuchAlgorithmException e) {
+            throw new GuacamoleException("Specified algorithm does not exist.", e);
+        }
+        catch (InvalidKeyException e) {
+            throw new GuacamoleException("Specified key is invalid.", e);
+        }
+        catch (InvalidKeySpecException e) {
+            throw new GuacamoleException("Invalid KeySpec initialization.", e);
+        }
+        catch (NoSuchPaddingException e) {
+            throw new GuacamoleException("No such padding exception.", e);
+        }
+
```

Comment:
And, you know how I said that the exceptions shouldn't kill the entire client, it should just result in the password not being available...well...I think all of this here killed that off...

Reply:
What do you mean by "killed that off"?

Reply:
When all of those exceptions were caught and dealt with directly in the AuthenticationProviderService class, the mehtod just returned null rather than throwing a GuacamoleException.  Returning null just means that the password is not added to the credential object, so it isn't available as a token later on inside the client.

Since creating a more generic type, here, I figured, while the behavior of just returning a null value and not killing authentication entirely is fine for the CAS extension, it may not be desirable across the board for whatever stuff may come later that might use this property, so instead of returning null, here, I'm throwing GuacamoleException.  This means that, if those exceptions are thrown while CAS is using this property, it'll kill CAS authentication if this Cipher property fails to parse for one of these reasons.  I consider that behavior less desirable than just not having the password available to the user, as authentication has technically still succeeded.

I suppose I could capture the GuacamoleException inside the CAS module when pulling this property and just return null instead of allowing the GuacamoleException to pass through...

Reply:
> ... so instead of returning null, here, I'm throwing GuacamoleException.

Sounds good. I'm always in favor of strict error handling, though [`GuacamoleServerException`](http://guacamole.incubator.apache.org/doc/guacamole-common/org/apache/guacamole/GuacamoleServerException.html) may be a better choice.

> This means that, if those exceptions are thrown while CAS is using this property, it'll kill CAS authentication if this Cipher property fails to parse for one of these reasons. I consider that behavior less desirable than just not having the password available to the user, as authentication has technically still succeeded.

Is it expected that this will occasionally fail under normal (not misconfigured) conditions? If not, I'd think the administrator would rather be able to rely on all-or-nothing, predictable behavior, particularly if relying on this feature.

> I suppose I could capture the GuacamoleException inside the CAS module when pulling this property and just return null instead of allowing the GuacamoleException to pass through...

That would depend on which is worse: having a partially-functional system, where explicitly-enabled functionality is mysteriously non-functional, or having a disabled system, where authentication fails entirely until everything which has been explicitly enabled is actually functional.

Reply:
> Is it expected that this will occasionally fail under normal (not misconfigured) conditions?

The more I think about it, the more I think this probably would not fail under anything but a misconfiguration.  In order to get ClearPass to work in CAS you have to generate public and private keys, and configure a service in CAS specifically to support the encryption of the stored password, so anyone who is setting this up also has to be setting up the CAS side of things, and if they get it wrong, well, it probably shouldn't work at all.  I'll change the exception to GuacamoleServerException and stick with failing authentication if this fails.

Reply:
So, I've redone most of this such that it throws the GuacamoleServerException.  There are two scenarios I can think of where having authentication succeed despite some error in the ClearPass decryption process would be desirable:
- If the credentials object is provided by the CAS server, but the Guacamole admin has not configured a private key, I think authentication should still succeed.  Since, in many organizations, SSO is run by someone different than a VDI/Desktop/RemoteAccess person, it's conceivable that the CAS server may provide something we choose not to consume, and that should not cause an error.
- Where the Guacamole admin has configured a PrivateKey, but CAS is not providing a value for the credential parameter.  Again, with the potential for CAS and Guacamole to be run by different admins/groups, or for different users within CAS to have different policies applied, it's conceivable that the GuacamoleAdmin configures a PrivateKey file for this purpose, but the attribute is blank/null.

Is my logic sound there?